Added correctness tests. Fixed some bugs.

jminor opened the pull request:
Hi, Thanks for making this open source! I'm not sure if you're looking for contributions, but I found a couple of bugs & fixed them.

Change:
Added correctness tests that verify board layout, simple moves, and a trivial checkmate-in-one-move puzzle for the AI.

Fixed a bug that caused the FiftyMove counter to only count moves by black.
Fixed a bug that caused the MoveCount to be initialized incorrectly from a FEN board.
Added some algebraic notation helper methods.

If you'd any changes, let me know.

3583Bytes commented:
Thank you for your changes and suggestions

If you change the 50 move count the way you suggested you would need 2 counters 1 for White and one for Black, else you would reach 50 after 25 moves.

jminor commented:
Oops, you're right, I messed that up. I've fixed it now.

I switched it to keep a half-move clock (to match the FEN notation) and comparing against 100 instead of 50 and included a test case to make sure it kicks in after the game has reached 50 full moves, not just 25.

To be double sure, I also verified this behavior against Stockfish: https://github.com/official-stockfish/Stockfish/blob/master/src/position.cpp#L1082
and Chess.js: https://github.com/jhlywa/chess.js/blob/master/chess.js#L1271

jminor commented:
Yay :)
